service/history/callback_queue_active_task_executor.go

```diff
@@ -0,0 +1,257 @@
+	ce, err := ms.GetCompletionEvent(ctx)
+	if err != nil {
+		return nil, err
+	}
+	switch ce.GetEventType() {
+	case enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_COMPLETED:
+		payloads := ce.GetWorkflowExecutionCompletedEventAttributes().GetResult().GetPayloads()
+		var payload *commonpb.Payload
+		if len(payloads) > 0 {
+			payload = payloads[0]
+		} else {
+			payload = &commonpb.Payload{}
+		}
+		completion, err := nexus.NewOperationCompletionSuccessful(payload, nexus.OperationCompletionSuccesfulOptions{
+			Serializer: t.payloadSerializer,
+		})
+		if err != nil {
+			return nil, serviceerror.NewInternal(fmt.Sprintf("failed to construct Nexus completion: %v", err))
+		}
+		return completion, nil
+	case enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_FAILED:
+		failure := commonnexus.APIFailureToNexusFailure(ce.GetWorkflowExecutionFailedEventAttributes().GetFailure())
+		return &nexus.OperationCompletionUnsuccessful{
+			State:   nexus.OperationStateFailed,
+			Failure: failure,
+		}, nil
+	case enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_CANCELED:
+		return &nexus.OperationCompletionUnsuccessful{
+			State:   nexus.OperationStateCanceled,
+			Failure: &nexus.Failure{Message: "operation canceled"},
+		}, nil
+	case enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_TERMINATED:
+		return &nexus.OperationCompletionUnsuccessful{
+			State:   nexus.OperationStateFailed,
+			Failure: &nexus.Failure{Message: "operation terminated"},
+		}, nil
+	case enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_TIMED_OUT:
+		return &nexus.OperationCompletionUnsuccessful{
+			State:   nexus.OperationStateFailed,
+			Failure: &nexus.Failure{Message: "operation exceeded internal timeout"},
+		}, nil
+	}
+	return nil, queues.NewUnprocessableTaskError(fmt.Sprintf("invalid workflow execution status: %v", ce.GetEventType()))
+}
+
+func (t *callbackQueueActiveTaskExecutor) processNexusCallbackTask(ctx context.Context, task *tasks.CallbackTask, url string, completion nexus.OperationCompletion) (retErr error) {
+	request, err := nexus.NewCompletionHTTPRequest(ctx, url, completion)
+	if err != nil {
+		return serviceerror.NewInternal(fmt.Sprintf("failed to construct Nexus request: %v", err))
+	}
+
+	caller := t.callerGetter(task.NamespaceID, task.DestinationAddress)
+	response, callErr := caller(request)
+	wfCtx, release, ms, err := t.getValidatedMutableStateForTask(
+		ctx, task, func(ms workflow.MutableState) error {
+			return t.validateCallbackTask(ms, task)
+		},
+	)
+	if err != nil {
+		return nil
+	}
+	defer func() { release(retErr) }()
+
+	// Callback is validated above and must be present in map.
+	callback := ms.GetExecutionInfo().Callbacks[task.CallbackID]
+	if err := t.applyCallbackResponse(ms, callback, response, callErr); err != nil {
+		return err
+	}
+	// Can't use UpdateWorkflowExecutionAsActive since it updates the current run and we are operating on closed
+	// workflows.
+	// When we support workflow-update callbacks, we'll need to revisit this code.
```

Comment:
Do we have a task for this? I think it's ok to leave this comment here as another reminder in case we forget, but I think a task is probably a better first line of defense.